clippy_utils/src/lib.rs

```diff
@@ -1224,6 +1224,57 @@ pub fn get_parent_as_impl(tcx: TyCtxt<'_>, id: HirId) -> Option<&Impl<'_>> {
     }
 }
 
+/// Remove blocks around an expression.
+///
+/// Ie. `x`, `{ x }` and `{{{{ x }}}}` all give `x`. `{ x; y }` and `{}` return
+/// themselves.
+pub fn peel_blocks<'a>(mut expr: &'a Expr<'a>) -> &'a Expr<'a> {
+    while let ExprKind::Block(
+        Block {
+            stmts: [],
+            expr: Some(inner),
+            rules: BlockCheckMode::DefaultBlock,
+            ..
+        },
+        _,
+    ) = expr.kind
+    {
+        expr = inner;
+    }
+    expr
+}
+
+/// Remove blocks around an expression or single statement.
+///
+/// Example: `x`, `{ x }`, `{{ x }}` or `{{ x; }}` all yield `x` but `{ x; y }` yields itself.
+pub fn peel_blocks_and_stmts<'a>(mut expr: &'a Expr<'a>) -> &'a Expr<'a> {
```

Comment:
I first reviewed the fallout fixes and after that this function. From reading the function name without looking at the function, I would have expected that this removes all blocks and also all statements, and returning the `expr` if available.

Maybe `peel_blocks_around_stmt`. Definitely the plural of `stmts` is confusing here.

Reply:
I was also wondering about it so I'm glad you raised this. How about `peel_stmt_blocks`? Or `peel_blocks_and_stmt`, `peel_blocks_with_stmt`.

Reply:
Of those three I like `peel_blocks_with_stmt` the most. I find it really hard to come up with a short and telling name here...

Reply:
Or `peel_blocks_from_stmt`?

Reply:
Ideally the name would convey "blocks having one expression with or without a statement" instead of "blocks with one statement". It is certainly tricky. I'm now leaning to `peel_blocks_with_stmt`.

Reply:
I think the name should better reflect the difference between this and the other until function. `peel_blocks_with_stmt` sounds to me like it would ignore all statements and only return the innermost expression while ignoring everything else. What do you think about `extract_expr_or_stmt` with the upper being called `extract_expr`? Or if we want to keep `peel` in the name, maybe `peel_all_blocks`?

I'm not quite sold on these names, but I feel like they capture the behavior a bit better :upside_down_face:

Reply:
It sounds like you had the same misconception as flip1995, unfortunately.

I think "block" is quite important - removing blocks is the primary function of the...function.

The word "peel" is consistent with some other methods like `Ty::peel_refs` and `Expr::peel_drop_temps`. It's not crucial, but I like how short it is.

Here is my thought process: to me the word "peel" means remove a layer and get *everything* inside of it. It doesn't mean remove a layer and then pick out some unspecified piece of what's inside. So I assume that `peel_blocks` returns everything inside the block, or else it will not "peel". And since it is `Expr -> Expr`, I assume it's looking for blocks containing a single expr. Now I admit that those are pretty hefty assumptions. But we have to choose *some* assumptions if we want a function name that is reasonably short. Documentation fills in the gaps. Honestly I think this is one of those cases where you're just not going know until you go read the docs.

`extract_expr` lends to the misconception you had even more strongly IMO, since "extract" conveys taking a part from the whole.

`peel_blocks` is already plural, so `peel_all_blocks` doesn't add any information.

So I'm currently liking `peel_blocks` and `peel_blocks_with_stmt`, but certainly open to discuss.

Reply:
> It sounds like you had the same misconception as flip1995, unfortunately.

I guessed as much from the previous discussion. However, I think that this shows that the original name was somewhat confusing.

Let's take `peel_blocks_with_stmt` for now, the more I think about it the more I like it :upside_down_face: